event.go

```diff
@@ -517,101 +711,239 @@ func checkID(id, kind string, sigil byte) (domain string, err error) {
 }
 
 // Origin returns the name of the server that sent the event
-func (e Event) Origin() ServerName { return e.fields.Origin }
+func (e *Event) Origin() ServerName {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Origin
+	case eventFormatV2Fields:
+		return fields.Origin
+	default:
+		panic(e.invalidFieldType())
+	}
+}
+
+func (e *Event) generateEventID() (eventID string, err error) {
+	switch e.roomVersion {
+	case RoomVersionV1, RoomVersionV2:
+		eventID = e.fields.(eventFormatV1Fields).EventID
+	case RoomVersionV3, RoomVersionV4, RoomVersionV5:
+		eventJSON := e.eventJSON
+		var reference EventReference
+		reference, err = referenceOfEvent(eventJSON, e.roomVersion)
+		if err != nil {
+			return
+		}
+		eventID = reference.EventID
+	default:
+		err = errors.New("gomatrixserverlib: unknown room version")
+	}
+	return
+}
 
 // EventID returns the event ID of the event.
-func (e Event) EventID() string {
-	return e.fields.EventID
+func (e *Event) EventID() string {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.EventID
+	case eventFormatV2Fields:
+		return fields.EventID
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Sender returns the user ID of the sender of the event.
-func (e Event) Sender() string {
-	return e.fields.Sender
+func (e *Event) Sender() string {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Sender
+	case eventFormatV2Fields:
+		return fields.Sender
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Type returns the type of the event.
-func (e Event) Type() string {
-	return e.fields.Type
+func (e *Event) Type() string {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Type
+	case eventFormatV2Fields:
+		return fields.Type
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // OriginServerTS returns the unix timestamp when this event was created on the origin server, with millisecond resolution.
-func (e Event) OriginServerTS() Timestamp {
-	return e.fields.OriginServerTS
+func (e *Event) OriginServerTS() Timestamp {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.OriginServerTS
+	case eventFormatV2Fields:
+		return fields.OriginServerTS
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Unsigned returns the object under the 'unsigned' key of the event.
-func (e Event) Unsigned() []byte {
-	return []byte(e.fields.Unsigned)
+func (e *Event) Unsigned() []byte {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Unsigned
+	case eventFormatV2Fields:
+		return fields.Unsigned
```

Comment:
All of this boilerplate feels unnecessary. It feels like you should have an interface so you don't need to constantly type switch..

Reply:
But then we just end up with two function implementations for each set of fields for each event format, so we don't gain on lines of code - it would probably look worse.